hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/HandlerBalancer.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.nio.tcp.handlermigration;
+
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.LoggingService;
+import com.hazelcast.nio.Connection;
+import com.hazelcast.nio.tcp.IOSelector;
+import com.hazelcast.nio.tcp.InSelectorImpl;
+import com.hazelcast.nio.tcp.MigratableHandler;
+import com.hazelcast.nio.tcp.OutSelectorImpl;
+import com.hazelcast.nio.tcp.ReadHandler;
+import com.hazelcast.nio.tcp.TcpIpConnection;
+import com.hazelcast.nio.tcp.WriteHandler;
+import com.hazelcast.util.EmptyStatement;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * It attempts to detect and fix a selector imbalance problem.
+ *
+ * Hazelcast by default uses 3 threads to read data from TCP connections and 3 threads to write data to connections.
+ * We have measured significant fluctuations of performance when the threads are not utilized equally.
+ *
+ * <code>com.hazelcast.nio.tcp.handlermigration.HandlerBalancer</code> tries to detect such situations and fix
+ * them by moving {@link com.hazelcast.nio.tcp.ReadHandler} and {@link com.hazelcast.nio.tcp.WriteHandler} between
+ * threads.
+ *
+ * It measures number of events serviced by each handler in a given interval and if imbalance is detected then it
+ * schedules handler migration to fix the situation. The exact migration strategy can be customized via
+ * {@link com.hazelcast.nio.tcp.handlermigration.MigrationStrategy}.
+ *
+ * Measuring interval can be customized via {@link com.hazelcast.instance.GroupProperties#HANDLER_MIGRATION_INTERVAL_SECONDS}
+ *
+ * It doesn't leverage {@link com.hazelcast.nio.ConnectionListener} capability
+ * provided by {@link com.hazelcast.nio.ConnectionManager} to observe connections as it has to be notified
+ * right after a physical TCP connection is created whilst <code>ConnectionListener</code> is notified only
+ * after a successful (Hazelcast) binding process.
+ *
+ *
+ */
+public class HandlerBalancer implements Runnable {
```

Comment:
Few suggestions:

Rename this class to IOBalancer.

Don't implement Runnable, but have an internal Thread implementation. Nobody needs to call run on your HandlerBalancer, so no need to expose it. I would totally internalize the thread like we do for a lot of other services. For example the SlowOperationDetector.